design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,477 @@
+With separated daemon key info tracking, the status will look like so:
+
+```yaml
+status:
+  # ...
+  cephx:
+    admin:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mon:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mgr:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    osd:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+Keeping track of each daemon type separately helps Rook ensure it won't re-rotate keys from earlier
+in the reconcile if the reconcile needs to restart later in the process. For example, there is no
+need to re-rotate admin/mon/mgr keys in the event the reconcile needs to restart in the middle of
+OSD updates. Because OSD updates can take quite a while, this case is likely to occur across Rook's
+large user base.
+
+### CSI example
+
+CSI keys are updated as part of a CephCluster reconcile.
+
+For a CephCluster reconcile:
+
+1. Ceph-CSI daemon keys (provisioners and node plugins) use `CephCluster.spec.security.cephx.csi`
+    (not `...daemon`) to determine if the reconcile should rotate.
+2. Assuming the rotation is indicated, rotate keys.
+3. Update the CephCluster status with the CSI key info
+
+    ```yaml
+    kind: CephCluster
+    # ...
+    status:
+      # ...
+      cephx:
+        # (status from CephCluster daemon rotations)
+        csi:
+          keyGeneration: 2 # e.g.
+          keyCephVersion: "20.2.1" # e.g.
+          priorKeyCount: 1 # e.g.
+    ```
+
+### CephFilesystem example
+
+Rook creates at least 2 MDSes for each CephFilesystem. This is a simple but non-trivial example of
+key rotation and an example showing rotation for a non-CephCluster resource.
+
+1. Because MDSes are daemons, the reconcile reads `CephCluster.spec.security.cephx.daemon` key
+    rotation configs to determine if the reconcile should rotate.
+2. When key rotation is indicated, it will rotate keys, first for MDS.a, then MDS.b, etc.
+3. Once reconciliation is nearing completion, CephFilesystem status is updated, as so:
+
+    ```yaml
+    status:
+      # ...
+      cephx:
+        daemon: # don't need this distinction, but keeps consistent status API and allows future change
+          keyGeneration: 2 # e.g.
+          keyCephVersion: "20.2.0" # e.g.
+    ```
+
+### CephRbdMirror example
+
+CephRbdMirror keys have both daemon and non-daemon keys. For daemon keys, rotation follows the
+    methods explained for CephFilesystem above.
+
+For the non-daemon (peer) key:
+
+1. Reconciliation reads `spec.security.cephx.peer` configs to determine if reconcile should rotate.
+2. When key rotation is indicated, the peer key is rotated.
+3. Once reconciliation is nearing completion, the status is updated.
+
+Because peer keys might be rotated locally and need an unknown amount of time for the peer to be
+updated, overlapping rotation will be used.
```

Comment:
@sp98 FYI we discussed in huddle that overlapping rotation will be good for CSI and mirror peer keys.